File: simple-chatbot/go/tools/tools.go
package tools

import (
	"errors"

	"github.com/firebase/genkit/go/ai"
	"github.com/firebase/genkit/go/genkit"
)

// TempConversionInput defines the input structure for the tool.
type TempConversionInput struct {
	Temperature float64 `json:"temperature"`
	From        string  `json:"from" jsonschema_description:"Unit to convert from (celsius or fahrenheit)"`
	To          string  `json:"to" jsonschema_description:"Unit to convert to (celsius or fahrenheit)"`
}

// DefineTempConversionTool defines the temperature conversion tool.
func DefineTempConversionTool(g *genkit.Genkit) ai.Tool {
	return genkit.DefineTool(
		g, "convertTemperature", "Converts temperature from one unit to another.",
		func(ctx *ai.ToolContext, input TempConversionInput) (float64, error) {
			if input.From == input.To {
				return input.Temperature, nil
			}
			if input.From == "celsius" && input.To == "fahrenheit" {
				return (input.Temperature*9)/5 + 32, nil
			}
			if input.From == "fahrenheit" && input.To == "celsius" {
				return ((input.Temperature - 32) * 5) / 9, nil
			}
			return 0, errors.New("invalid temperature conversion units")
		},
	)
}

File: simple-chatbot/src/tools.ts
import { GenkitBeta, z } from 'genkit/beta';

export const toolNames = ['temperatureConverter'];

export function defineTools(ai: GenkitBeta) {
  ai.defineTool(
    {
      name: 'temperatureConverter',
      description: 'Converts temperature between Celsius and Fahrenheit.',
      inputSchema: z.object({
        temperature: z.number(),
        from: z.enum(['C', 'F']),
        to: z.enum(['C', 'F']),
      }),
      outputSchema: z.number(),
    },
    async ({ temperature, from, to }) => {
      if (from === 'C' && to === 'F') {
        return (temperature * 9) / 5 + 32;
      }
      if (from === 'F' && to === 'C') {
        return ((temperature - 32) * 5) / 9;
      }
      throw new Error('Invalid conversion units.');
    }
  );
}
